src/client.ts

```diff
@@ -1,149 +1,224 @@
-/**
- * An APIClient maintains authentication state and provides an interface to
- * third party data APIs.
- *
- * It is recommended that integrations wrap provider data APIs to provide a
- * place to handle error responses and implement common patterns for iterating
- * resources.
- */
 export class APIClient {
-  constructor(readonly config: IntegrationConfig) {}
+  private readonly clientId: string;
+  private readonly clientSecret: string;
+  private readonly accountId: string;
+  /*
+    Explanation: the reason this accessToken variable is static class variable 
+    is that we don't want each step function's run to request the same (new) token
+    every time. This way, once accessToken value is fetched, it's shared between
+    all the step functions. 
+  */
+  private static accessToken: string;
+
+  constructor(readonly config: CalculatedIntegrationConfig) {
+    this.accountId = config.accountId;
+    this.clientId = config.clientId;
+    this.clientSecret = config.clientSecret;
+  }
 
-  public async verifyAuthentication(): Promise<void> {
-    // TODO make the most light-weight request possible to validate
-    // authentication works with the provided credentials, throw an err if
-    // authentication fails
-    const request = new Promise((resolve, reject) => {
-      http.get(
-        {
-          hostname: 'localhost',
-          port: 443,
-          path: '/api/v1/some/endpoint?limit=1',
-          agent: false,
-          timeout: 10,
-        },
-        (res) => {
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
+  private withBaseUri(path: string): string {
+    return `https://api.malwarebytes.com/${path}`;
+  }
+
+  private async request(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    body?: any,
+  ): Promise<Response> {
+    if (!APIClient.accessToken) {
+      await this.getAuthenticationToken();
+    }
+
+    return await fetch(uri, {
+      method,
+      headers: {
+        Authorization: `Bearer ${APIClient.accessToken}`,
+        AccountId: this.accountId,
+        'Content-Type': 'application/json',
+      },
+      body: body || null,
+    });
+  }
+
+  private async makePaginatedRequest<K extends string, V>(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    resource: K,
+    body?: any,
+  ): Promise<V[]> {
+    const results: V[] = [];
+    let nextCursor = '';
+
+    do {
+      const response = await this.request(
+        uri,
+        method,
+        JSON.stringify({
+          ...body,
+          page_size: 10,
+          next_cursor: nextCursor,
+        }),
       );
+
+      const responseJson: PaginatedDataResponse<K, V[]> = await response.json();
+
+      if (responseJson) {
+        nextCursor = responseJson.next_cursor;
+        results.push(...responseJson[resource]);
+      }
+    } while (nextCursor && nextCursor.length > 0);
```

Comment:
This `while` condition is interesting... this seems to suggest that when pagination is complete, the `next_cursor` property is _either_ `undefined` or `''`. Is that true?

Reply:
Yes, you're right - From what I remember (I don't have code running/next to me at this moment sadly) it was '' (empty string) when the pagination is complete. I'll update this answer in case I'm wrong at the moment!

Reply:
Just to confirm this - if there are no further pages, next_cursor is "" (empty string). If there is a next page, the next_cursor has that identifying value in it.

Reply:
Ok. Can you confirm the following, just so I'm 100% clear:

Simply using `while (nextCursor.length > 0)` will _not_ work here, because there is a case that `nextCursor` is undefined.

If that's the case, I would like to propose the following, because using the `.length` property is a little ambiguous to me (e.g. is it an array? is it a string?)

```suggestion
    } while (nextCursor && nextCursor !== '');
```